Explicit copyout changes

bsdjhb opened the pull request:
This is the rest of the changes to use explicit user caps.  After this I will start on the cleanups to remove *_c and retire the !EXPLICIT_USER_ACCESS code, etc.

If this is too big of a chunk I can split this up for review.

qwattash commented on sys/cam/cam_periph.c:
```diff
@@ -932,7 +934,7 @@ cam_periph_mapmem(union ccb *ccb, struct cam_periph_map_info *mapinfo,
 
 		/* Save the user's data address. */
-		mapinfo->orig[i] = *data_ptrs[i];
+		mapinfo->orig[i] = __USER_CAP(*data_ptrs[i], lengths[i]);
```
I'm not super familiar with this part of the code, however it seems that `data_ptrs[]` holds `uint8_t*` and `struct ccb_dev_match` (for example) contains various pointers. I believe that I will have to do some `uint64_t` or `__ptr64` annotation for these to work in the purecap kernel. Do you know which userspace tool uses these ioctls?

bsdjhb replied:
I don't know, and it could be out-of-tree things like smartctl, etc.  The pass(4) driver used for SCSI pass through commands uses this interface.  camcontrol might be another one.  However, the goal here is not to let freebsd64 versions work, but to let purecap versions work.  Perhaps data_ptrs[] array needs to be an explicit array of capabilities instead.

Ah, yes.  I need to instead go splat a bunch of __kerncap annotations in `union ccb` and rework some of this to assume user capabilities instead.  If we need freebsd64 support we will have to do that separately.

bsdjhb replied:
I've reworked `union ccb` to actually be purecap friendly which removed all the `__USER_CAP` uses in this file.

qwattash commented on sys/cam/ctl/ctl_frontend_ioctl.c:
```diff
@@ -367,7 +369,7 @@ ctl_ioctl_do_datamove(struct ctl_scsiio *ctsio)
 							   M_WAITOK);
 		ext_sglist_malloced = 1;
-		if (copyin(ctsio->ext_data_ptr, ext_sglist, ext_sglen) != 0) {
+		if (copyin(ctsio->ext_data_uptr, ext_sglist, ext_sglen) != 0) {
```
It is unclear to me what sets `ext_data_uptr`, do we need a corresponding `__USER_CAP` conversion from it somewhere, or we assume that this will not work unless the userspace is purecap?

bsdjhb replied:
Userland code that sets `ext_data_ptr` actually sets `ext_data_uptr`.  The union only exists in the kernel, and in purecap userland `ext_data_ptr` is a capability.

qwattash replied:
Makes sense now, thanks

qwattash commented on sys/cam/scsi/scsi_pass.c:
```diff
@@ -1238,7 +1242,9 @@ passcopysglist(struct cam_periph *periph, struct pass_io_req *io_req,
 					  "error %d\n", __func__, len_to_copy,
-					  kern_ptr, user_ptr, error);
+					  kern_ptr,
+					  (void *)(__cheri_addr uintptr_t)user_ptr,
```
Perhaps we should add a capability format specifier to printf, these casts are quite common.

bsdjhb replied:
The problem is you can't pass a capability via var-args in hybrid.  It's not implemented in the ABI for either MIPS or RISC-V.  @jrtc27 had some ideas such as passing a pointer to the capability instead as the ABI to use implicitly, but that needs some toolchain work.

qwattash replied:
Right, I forgot that was the reason. I don't have a strong opinion on this, I'd say it depends on how long we plan to support hybrid. It wouldn't reduce the diff too much as we would need to change the format string anyway.

qwattash commented on sys/cam/scsi/scsi_pass.c:
```diff
@@ -1219,12 +1221,14 @@ passcopysglist(struct cam_periph *periph, struct pass_io_req *io_req,
 		    kern_sglist[j].ds_len - kern_watermark);
 
-		user_ptr = (uint8_t *)(uintptr_t)user_sglist[i].ds_addr;
+		user_ptr = __USER_CAP((uint8_t *)(uintptr_t)user_sglist[i].ds_addr,
```
Now `__USER_CAP` accepts `uint64_t` without need to cast. I feel it is somewhat redundant as the macro already makes it explicit that the result will be a pointer.
As an aside (and as a note to self), going back to our last conversation about `bus_addr_t`, it seems that there are different interpretations of `bus_addr_t` across architectures on whether it holds a physical address, an opaque value of some kind or a virtual address. This seems to be one of those places that assumes it can hold a virtual address.

bsdjhb replied:
Here I just kept the existing code, though perhaps I need to adjust this as part of my earlier comment to permit `ds_addr` to be a capability.  I'll know more about how I'd like to handle that (possibly a union around `ds_addr`) when I look at this again.

bsdjhb replied:
This no longer uses `__USER_CAP`.

qwattash commented on sys/compat/linuxkpi/common/src/linux_compat.c:
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(kaddr, uaddr, len);
+		memcpy(kaddr, (void *)(__cheri_addr uintptr_t)uaddr, len);
```
This would fail in the purecap kernel. Should probably use `(__cheri_fromcap void *)uaddr` that keeps `uaddr` from becoming untagged in pruecap mode.

qwattash replied:
If it has not changed or has some special meaning with `uintptr_t`, `__cheri_addr` should not be ignored in purecap as I recall explicitly using it some places.

bsdjhb replied:
But uintptr_t is uintcap_t in purecap, not vaddr_t.  __cheri_addr for a vaddr_t cast makes sense still, but I would expect a uintcap_t cast of a capability pointer to always be a no-op since you aren't losing bits unlike uintptr_t in hybrid (which is vaddr_t) where you are only getting 64 bits as a result of the cast.

qwattash replied:
I just checked what the current compiler does and it crashes when casting a pointer to `(__cheri_addr __intcap_t)`, I do agree that it should mirror `__cheri_to/fromcap` and be a no-op in this case, otherwise it is confusing. I'm opening a clang issue about this.

bsdjhb replied:
This now uses `__cheri_fromcap`

qwattash commented on sys/compat/linuxkpi/common/src/linux_compat.c:
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
 		if (uaddr == NULL)
 			return (-EFAULT);
-		memcpy(uaddr, kaddr, len);
+		memcpy((void *)(__cheri_addr uintptr_t)uaddr, kaddr, len);
```
Same as above

qwattash commented on sys/compat/linuxkpi/common/src/linux_compat.c:
```diff
@@ -824,44 +826,44 @@ linux_remap_address(void **uaddr, size_t len)
 
 	/* make sure uaddr is aligned before going into the fast loop */
-	while (((uintptr_t)uaddr & 7) != 0 && len > 7) {
+	while (((__cheri_addr uintptr_t)uaddr & 7) != 0 && len > 7) {
```
Probably a good place for `__builtin_is_aligned()`. I have an `is_aligned()` macro in the purecap kernel branch that I should probably have merged sooner. We could use the builtin here and then move it to the macro when we merge it.

arichardson replied:
Do we have wrapper macros for __builtin_is_aligned? I think we changed roundup2 and rounddown2 to use the builtins but not sure about this one.

bsdjhb replied:
We didn't change roundup/rounddown because they now DTRT as-is.  We had explicitly used the builtin in some places that might have gone back to using round* to minimize diffs with upstream.

bsdjhb replied:
I did end up using `__builtin_is_aligned` here.

qwattash commented on sys/contrib/ipfilter/netinet/fil.c:
```diff
@@ -8009,7 +8024,6 @@ ipf_getnextrule(softc, t, ptr)
 		next = &zero;
 		t->ipt_data = NULL;
 	}
-	it.iri_rule = predict ? next : NULL;
```
Why was this line deleted?

bsdjhb replied:
It's dead code that isn't used, and it's assigning a kernel pointer (vaddr) to a user pointer (capability) so fails to compile.

qwattash commented on sys/contrib/ipfilter/netinet/ip_fil.h:
```diff
@@ -1321,7 +1321,7 @@ typedef	int	(*ipftq_delete_fn_t)(struct ipf_main_softc_s *, void *);
 	u_32_t		ipfo_rev;	/* IPFilter version number */
 	u_32_t		ipfo_size;	/* size of object at ipfo_ptr */
-	void		*ipfo_ptr;	/* pointer to object */
+	void * __kerncap ipfo_ptr;	/* pointer to object */
```
nit. keep indentation consistent with the rest of the file
```suggestion
	void			* __kerncap ipfo_ptr;	/* pointer to object */
```

bsdjhb replied:
@brooksdavis has previously stated he prefers this version to avoid reindenting the rest of the line as much (e.g. the comment in this case still lines up with the other comments)

qwattash commented on sys/dev/mfi/mfi_ioctl.h:
```diff
@@ -57,13 +57,13 @@ union mfi_statrequest {
 union mfi_sense_ptr {
 	uint8_t		sense_ptr_data[MAX_SPACE_FOR_SENSE_PTR];
-	void 		*user_space;
+	void * __kerncap user_space;
```
nit. Keep indentation consistent with rest of file
```suggestion
	void			* __kerncap user_space;
```

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -592,12 +594,12 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 					    TI_SWAP_NTOH);
 					TI_UNLOCK(sc);
-					bcopy(&sc->ti_membuf2[segresid], ptr,
+					bcopy(&sc->ti_membuf2[segresid], (void *)(__cheri_addr uintptr_t)ptr,
```
`(void *)(__cheri_addr uintptr_t)ptr` will strip the tag in the purecap kernel.
```suggestion
					bcopy(&sc->ti_membuf2[segresid], (__cheri_fromcap void *)ptr,
```

arichardson replied:
Do we really want this to be kernel-ddc relative?

bsdjhb replied:
Well, making it a `void *` in hybrid is going to be kernel DDC relative anytime it is dereferenced.  Perhaps it should use fromcap?  I'd have to fix the other end of it to use tocap though instead of whatever it does now perhaps to stuff a kernel pointer into the capability to pass into this.

bsdjhb replied:
Ended up using `__cheri_fromcap`

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -592,12 +594,12 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 					first_pass = 0;
 				} else
-					ti_bcopy_swap(sc->ti_membuf, ptr,
+					ti_bcopy_swap(sc->ti_membuf, (void *)(__cheri_addr uintptr_t)ptr,
```
```suggestion
					ti_bcopy_swap(sc->ti_membuf, (__cheri_fromcap void *)ptr,
```

arichardson replied:
Same question here. I know in purecap __cheri_fromcap will do the right thing since it just returns the capability unchanged but in hybrid it will be relative to kernel ddc.

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -609,7 +611,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				    segsize, TI_SWAP_HTON);
 			} else
-				ti_bcopy_swap(ptr, sc->ti_membuf, segsize,
+				ti_bcopy_swap((void *)(__cheri_addr uintptr_t)ptr, sc->ti_membuf, segsize,
```
```suggestion
				ti_bcopy_swap((__cheri_fromcap void *)ptr, sc->ti_membuf, segsize,
```

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -659,7 +661,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				TI_LOCK(sc);
 			} else
-				bcopy(&tmpval2, ptr, resid);
+				bcopy(&tmpval2, (void *)(__cheri_addr uintptr_t)ptr, resid);
```
```suggestion
				bcopy(&tmpval2, (__cheri_fromcap void *)ptr, resid);
```

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -677,7 +679,7 @@ ti_copy_mem(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				TI_LOCK(sc);
 			} else
-				bcopy(ptr, &tmpval2, resid);
+				bcopy((void *)(__cheri_addr uintptr_t)ptr, &tmpval2, resid);
```
```suggestion
				bcopy((__cheri_fromcap void *)ptr, &tmpval2, resid);
```

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -765,12 +767,12 @@ ti_copy_scratch(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				copyout(&tmpval, ptr, 4);
 			else
-				bcopy(&tmpval, ptr, 4);
+				bcopy(&tmpval, (void *)(__cheri_addr uintptr_t)ptr, 4);
```
```suggestion
				bcopy(&tmpval, (__cheri_fromcap void *)ptr, 4);
```

qwattash commented on sys/dev/ti/if_ti.c:
```diff
@@ -765,12 +767,12 @@ ti_copy_scratch(struct ti_softc *sc, uint32_t tigon_addr, uint32_t len,
 				copyin(ptr, &tmpval2, 4);
 			else
-				bcopy(ptr, &tmpval2, 4);
+				bcopy((void *)(__cheri_addr uintptr_t)ptr, &tmpval2, 4);
```
```suggestion
				bcopy((__cheri_fromcap void *)ptr, &tmpval2, 4);
```

brooksdavis commented on sys/net80211/ieee80211_ioctl.c:
```diff
@@ -226,7 +228,7 @@ ieee80211_ioctl_getstastats(struct ieee80211vap *vap, struct ieee80211req *ireq)
 		ireq->i_len = sizeof(struct ieee80211req_sta_stats);
 	/* NB: copy out only the statistics */
-	error = copyout(&ni->ni_stats, (uint8_t *) ireq->i_data + off,
+	error = copyout(&ni->ni_stats, (uint8_t * __capability) ireq->i_data + off,
```
```suggestion
	error = copyout(&ni->ni_stats, (uint8_t * __capability)ireq->i_data + off,
```

brooksdavis commented on sys/net80211/ieee80211_ioctl.c:
```diff
@@ -492,7 +494,8 @@ getstainfo_common(struct ieee80211vap *vap, struct ieee80211req *ireq,
 			get_sta_info(&req, ni);
 		ireq->i_len = space - req.space;
-		error = copyout(p, (uint8_t *) ireq->i_data+off, ireq->i_len);
+		error = copyout(p, (uint8_t * __capability) ireq->i_data + off,
```
```suggestion
		error = copyout(p, (uint8_t * __capability)ireq->i_data + off,
```

brooksdavis commented on sys/dev/usb/usb_ioctl.h:
```diff
@@ -185,8 +185,8 @@ struct usb_fs_endpoint {
 #else
-	void  **ppBuffer;		/* pointer to userland buffers */
-	uint32_t *pLength;		/* pointer to frame lengths, updated
+	void  * __kerncap * __kerncap ppBuffer;	/* pointer to userland buffers */
+	uint32_t  * __kerncap pLength;	/* pointer to frame lengths, updated
```
```suggestion
	void * __kerncap * __kerncap ppBuffer;	/* pointer to userland buffers */
	uint32_t * __kerncap pLength;	/* pointer to frame lengths, updated
```

brooksdavis commented on sys/net/if_lagg.h:
```diff
@@ -116,7 +116,7 @@ struct lagg_reqall {
 
 	size_t			ra_size;		/* size of buffer */
-	struct lagg_reqport	*ra_port;		/* allocated buffer */
+	struct lagg_reqport	* __kerncap ra_port;	/* allocated buffer */
```
```suggestion
	struct lagg_reqport * __kerncap ra_port;	/* allocated buffer */
```

brooksdavis commented on sys/contrib/ipfilter/netinet/ip_fil.h:
```diff
@@ -1443,7 +1443,7 @@ typedef	struct	ipfruleiter {
 	int		iri_nrules;
 	int		iri_v;		/* No longer used (compatibility) */
-	frentry_t	*iri_rule;
+	frentry_t	* __kerncap iri_rule;
```
```suggestion
	frentry_t * __kerncap iri_rule;
```

brooksdavis commented on sys/contrib/ipfilter/netinet/ip_fil.h:
```diff
@@ -1458,7 +1458,7 @@ typedef	struct	ipfruleiter {
 	int	igi_type;
 	int	igi_nitems;
-	void	*igi_data;
+	void	* __kerncap igi_data;
```
```suggestion
	void * __kerncap igi_data;
```

brooksdavis commented on sys/contrib/ipfilter/netinet/ip_fil.h:
```diff
@@ -1473,7 +1473,7 @@ typedef	struct	ipfgeniter {
 typedef	struct	ipftable {
 	int	ita_type;
-	void	*ita_table;
+	void	* __kerncap ita_table;
```
```suggestion
	void * __kerncap ita_table;
```

brooksdavis commented on sys/contrib/ipfilter/netinet/ip_lookup.h:
```diff
@@ -39,7 +39,7 @@ typedef	struct	iplookupop	{
 	char	iplo_name[FR_GROUPLEN];
 	size_t	iplo_size;	/* sizeof struct at iplo_struct */
-	void	*iplo_struct;
+	void	* __kerncap iplo_struct;
```
```suggestion
	void * __kerncap iplo_struct;
```

brooksdavis commented on sys/contrib/ipfilter/netinet/ip_lookup.h:
```diff
@@ -84,7 +84,7 @@ typedef	struct	ipflookupiter	{
 	iplookupiterkey_t	ili_lkey;
 	char			ili_name[FR_GROUPLEN];
-	void			*ili_data;
+	void			* __kerncap ili_data;
```
```suggestion
	void * __kerncap	ili_data;
```

brooksdavis commented on sys/dev/iscsi/iscsi_ioctl.h:
```diff
@@ -220,7 +220,7 @@ struct iscsi_session_remove {
 struct iscsi_session_list {
 	unsigned int			isl_nentries;
-	struct iscsi_session_state	*isl_pstates;
+	struct iscsi_session_state	* __kerncap isl_pstates;
```
```suggestion
	struct iscsi_session_state * __kerncap isl_pstates;
```

brooksdavis commented on sys/dev/iscsi_initiator/iscsi.h:
```diff
@@ -422,11 +422,21 @@ typedef struct {
+	     ahs_t * __capability ahs_uaddr;
+#endif
+	     ahs_t	*ahs_addr;
+     };
```
I wonder if judicious use of macros could make these sorts of changes smaller without making them too confusing.

bsdjhb replied:
I'm not sure.  In some cases I wonder if I I couldn't get by with making them __kerncap and just fixing some in-kernel uses to use __fromcap and __tocap casts, but usually it seemed like overloaded pointers were used too widely for that.

bsdjhb commented:
Ok, I've reworked some bits of this.  The `(__cheri_addr uintptr_t)` casts should generally be replaced with `__cheri_fromcap` and `__cheri_tocap` pointer casts.  I've also marked up `union ccb` so it actually works for purecap and updated CAM to cope.  This meant that the `__USER_CTOPTR` thing is gone now.  I've also added a union around `ds_addr` in `bus_dma_segment_t` to hold pointers since CAM does that.  As a sanity check, the only places that now use `__USER_CAP*` are in freebsd32 and freebsd64 compat, which is the only places that should be using those.

bsdjhb commented on sys/sys/systm.h:
```diff
@@ -207,10 +207,24 @@ void	kassert_panic(const char *fmt, ...)  __printflike(1, 2);
+	((void *)(__builtin_cheri_tag_get(cap) ?			\
+	    (uintptr_t)((char * __capability)(cap) -			\
+		(char * __capability)(base)) :				\
+	    (__cheri_addr uintptr_t)(cap)))
```
I ended up dropping these macros entirely after the `union ccb` changes.
